File: src/main/java/rip/hippo/mosey/asm/ClassHierarchy.java
/*
 * Mosey is a free and open source java bytecode obfuscator.
 *     Copyright (C) 2020  Hippo
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package rip.hippo.mosey.asm;


import org.tinylog.Logger;
import rip.hippo.mosey.asm.wrapper.ClassWrapper;

import java.util.*;

/**
 * @author Hippo
 * @version 1.0.0, 6/23/20
 * @since 1.0.0
 */
public enum ClassHierarchy {
    ;
    private static final Map<String, ClassWrapper> CLASS_LOOKUP_MAP = new HashMap<>();
    private static final Map<ClassWrapper, ClassWrapper> CLASS_HIERARCHY_MAP = new HashMap<>();
    
    public static void registerClass(ClassWrapper classWrapper) {
        CLASS_LOOKUP_MAP.put(classWrapper.getName(), classWrapper);
    }
    
    public static ClassWrapper lookup(String name) {
        return CLASS_LOOKUP_MAP.get(name);
    }

    public static void registerAncestors() {
        for (ClassWrapper classWrapper : CLASS_LOOKUP_MAP.values()) {
            setSuperClass(classWrapper, classWrapper.getSuperName());
        }
    }
    
    private static void setSuperClass(ClassWrapper classWrapper, String name) {
        ClassWrapper lookup = lookup(name);
        if(lookup == null) {
            Logger.warn(String.format("Class %s can't be found, ensure all libraries are loaded, ignoring.", name));
            return;
        }
        CLASS_HIERARCHY_MAP.put(classWrapper, lookup);
    }

    public static ClassWrapper getSuperClass(ClassWrapper classWrapper) {
        return CLASS_HIERARCHY_MAP.get(classWrapper);
    }
}

File: src/main/scala/rip/hippo/mosey/asm/ClassHierarchy.scala
package rip.hippo.mosey.asm
import rip.hippo.mosey.asm.wrapper.ClassWrapper
import rip.hippo.mosey.logger.Logger

import scala.collection.mutable

/**
 * @author Hippo
 * @version 1.0.0, 7/8/20
 * @since 1.0.0
 */
object ClassHierarchy {
  private val classLookup: mutable.Map[String, ClassWrapper] = mutable.Map()
  private val classHierarchy: mutable.Map[ClassWrapper, ClassWrapper] = mutable.Map()

  def registerClass(classWrapper: ClassWrapper): Unit =
    classLookup += (classWrapper.getName -> classWrapper)

  def lookup(name: String): Option[ClassWrapper] =
    classLookup.get(name)

  def registerSuperclasses: Unit =
    classLookup.values.foreach(classWrapper => setSuperClass(classWrapper, classWrapper.getSuperName))

  def setSuperClass(classWrapper: ClassWrapper, name: String): Unit =
    lookup(name) match {
      case None => Logger.warn(String.format("Class %s can't be found, ensure all libraries are loaded, ignoring.", name).toString)
      case Some(value) => classHierarchy += (classWrapper -> value)
    }

  def getSuperClass(classWrapper: ClassWrapper): ClassWrapper =
    classHierarchy(classWrapper)
}
